File: backend/others/SegmentLinker.py
from pymongo import MongoClient
import pprint
from Areas import Areas

db = MongoClient("localhost", 27017).pruebaDatosGeograficos

debug = False
debug2 = False


class SegmentLinker(object):

	# Devuelve el id del segmento más cercano a unas coordenadas
	def nearestSegment(self, longitud, latitud):
		segments = []
		if debug:
			print(
				"################################################################################")
		for doc in db.segmentos.find({"segment": {"$nearSphere": [longitud, latitud]}}).limit(2):
			a = Areas()
			rectangleVertices = a.buildRectangle(
				doc["segment"]["coordinates"][0], doc["segment"]["coordinates"][1])
			if debug:
				print("Longitud y latitud: ", longitud, ", ", latitud)
			if debug:
				print("Las coordenadas del segmento examinando: ",
					  doc["segment"]["coordinates"][0], ", ", doc["segment"]["coordinates"][1])
			if debug:
				print("Los vértices del rectángulo son: ", rectangleVertices)
			rectangleArea = a.calculateRectangleArea(
				rectangleVertices[0], rectangleVertices[1], rectangleVertices[2], rectangleVertices[3])
			if debug:
				print("El área del rectángulo es: ", rectangleArea)
			point = [longitud, latitud]
			triangleArea1 = a.calculateTriangleArea(
				point, rectangleVertices[0], rectangleVertices[1])
			triangleArea2 = a.calculateTriangleArea(
				point, rectangleVertices[1], rectangleVertices[2])
			triangleArea3 = a.calculateTriangleArea(
				point, rectangleVertices[2], rectangleVertices[3])
			triangleArea4 = a.calculateTriangleArea(
				point, rectangleVertices[3], rectangleVertices[0])
			sumTriangleAreas = triangleArea1+triangleArea2+triangleArea3+triangleArea4
			if debug:
				print("El sumatorio de los triángulos es: ", sumTriangleAreas)
			if rectangleArea <= (sumTriangleAreas+0.0000000001) and rectangleArea >= (sumTriangleAreas-0.0000000001):
				if debug:
					print("El codigo way es: ", doc["way"])
				segments.append(doc["way"])
				if debug:
					print(
						"-------------------------------------------------------------------------------")
		if debug:
			print(
				"################################################################################")
		return segments

	# Dados dos pares de coordenadas devuelve el segmento más cercano a ambos
	# Se tiene en cuenta que el movimiento es de las primeras coordenadas a las segundas para casos de conflicto
	def segmentByEmotion(self, longitud1, latitud1, longitud2, latitud2):
		segment1 = self.nearestSegment(longitud1, latitud1)
		segment2 = self.nearestSegment(longitud2, latitud2)

		if debug2:
			print(segment1)
		if debug2:
			print(segment2)

		# Ambas coordenadas tienen segments cerca
		if len(segment1) != 0 and len(segment2) != 0:
			if debug2:
				print(11111111111111111111111111111)
			# Ambas coordenadas corresponden a un segmento sin haber conflicto
			# Ambas coordenadas corresponden a la misma intersección
			if segment1 == segment2:
				return segment1[0]
			# La primera coordenada está en una intersección pero la segunda no
			elif len(segment1) > 1 and len(segment2) == 1:
				return segment2[0]
			# La segunda coordenada está en una intersección pero la primera no
			elif len(segment1) == 1 and len(segment2) > 1:
				return segment1[0]
			# Ambas coordenadas están en intersecciones pero no a la misma intersección
			elif len(segment1) > 1 and len(segment2) > 1 and segment1 != segment2:
				if segment1[0] == segment2[0]:
					return segment1[0]
				else:
					return segment1[1]
		# El primer par de coordenadas tiene algún segmento cerca pero el segundo no
		elif len(segment1) != 0:
			if debug2:
				print(22222222222222222222222222)
			return segment1[0]
		# El segundo par de coordenadas tiene algún segmento cerca pero el primero no
		else:
			if debug2:
				print(3333333333333333333333333333333)
			return segment2[0]


# a=SegmentLinker()
# a.nearestSegment(10,5)
#a.nearestSegment(41.652116, -0.880042);

File: backend/administration_app/spring-boot-quartz-demo-master/src/main/java/com/demo/service/SegmentLinker.java
package com.demo.service;

import java.util.ArrayList;

import com.demo.entity.Position;
import com.demo.entity.Segment;
import com.demo.repository.SegmentRepository;

import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Service;

@Service
public class SegmentLinker {

	private boolean debug = false;

	private SegmentRepository segmentRepository;

	public SegmentLinker(SegmentRepository segmentRepository) {
		this.segmentRepository = segmentRepository;
	}

	private ArrayList<Segment> closestSegment(Position point) {
		ArrayList<Segment> segments = new ArrayList<Segment>();
		Areas a = new Areas();
		double rectangleArea = 0;
		double triangleArea1 = 0;
		double triangleArea2 = 0;
		double triangleArea3 = 0;
		double triangleArea4 = 0;
		double sumTriangleAreas = 0;

		ArrayList<Segment> segmentsFound = segmentRepository.findTop2ByCoordinates(point.getLng(), point.getLat(),
				new PageRequest(0, 2));
		for (Segment s : segmentsFound) {
			if (debug)
				System.out.println("-----------------------------------------");
			if (debug)
				System.out.println("Uno de los segmentos encontrados: " + s.toString());
			ArrayList<Position> rectangleVertices = a.buildRectangle(s.getPoint1(), s.getPoint2());
			rectangleArea = a.calculateRectangleArea(rectangleVertices.get(0), rectangleVertices.get(1),
					rectangleVertices.get(2), rectangleVertices.get(3));
			triangleArea1 = a.calculateTriangleArea(point, rectangleVertices.get(0), rectangleVertices.get(1));
			triangleArea2 = a.calculateTriangleArea(point, rectangleVertices.get(1), rectangleVertices.get(2));
			triangleArea3 = a.calculateTriangleArea(point, rectangleVertices.get(2), rectangleVertices.get(3));
			triangleArea4 = a.calculateTriangleArea(point, rectangleVertices.get(3), rectangleVertices.get(0));
			sumTriangleAreas = triangleArea1 + triangleArea2 + triangleArea3 + triangleArea4;
			if (rectangleArea <= sumTriangleAreas + 0.0000000001 && rectangleArea >= sumTriangleAreas - 0.0000000001) {
				segments.add(s);
			}
		}

		return segments;
	}

	public Segment segmentByEmotion(Position point1, Position point2) {
		ArrayList<Segment> list1 = closestSegment(point1);
		ArrayList<Segment> list2 = closestSegment(point2);

		// Ambas coordenadas tienen segments cerca
		if (list1.size() != 0 && list2.size() != 0) {
			// Ambas coordenadas corresponden a un segmento sin haber conflicto
			// Ambas coordenadas corresponden a la misma intersección
			if (list1.equals(list2)) {
				return list1.get(0);
			}
			// La primera coordenada está en una intersección pero la segunda no
			else if (list1.size() > 1 && list2.size() == 1) {
				return list2.get(0);
			}
			// La segunda coordenada está en una intersección pero la primera no
			else if (list1.size() == 1 && list2.size() > 1) {
				return list1.get(0);
			} else if (list1.size() > 1 && list2.size() > 1 && !list1.equals(list2)) {
				if (list1.get(0).equals(list2.get(0))) {
					return list1.get(0);
				} else {
					return list1.get(1);
				}
			}
		}
		// El primer par de coordenadas tiene algún segmento cerca pero el segundo no
		else if (list1.size() != 0) {
			return list1.get(0);
		}
		// El segundo par de coordenadas tiene algún segmento cerca pero el primero no
		else {
			return list2.get(0);
		}
		return null;
	}
}